File: datasets/Python/Claude/Algorithm/hard5.py
def solve():
    MOD = 10**9 + 7
    
    # Read input
    n, m = map(int, input().split())
    
    # Available values: 1, 2, ..., 2^m - 1
    max_val = (1 << m) - 1  # 2^m - 1
    
    if n > max_val:
        print(0)
        return
    
    # We need to count ordered arrangements (permutations) of n distinct values
    # from {1, 2, ..., 2^m - 1} where first player wins (Nim-sum ≠ 0)
    
    # Mathematical approach:
    # 1. Total permutations = P(2^m - 1, n) = (2^m - 1)! / (2^m - 1 - n)!
    # 2. Subtract permutations where Nim-sum = 0
    
    # For Nim-sum = 0: we need sets of n values whose XOR is 0
    # Key insight: among all possible combinations of n distinct values,
    # exactly C(2^(m-1) - 1, n-1) combinations have XOR = 0 (when n > 1)
    # Each such combination contributes n! permutations
    
    # Precompute factorials (limit to reasonable size)
    max_fact = min(max_val + 1, 2*10**6)
    fact = [1] * max_fact
    for i in range(1, max_fact):
        fact[i] = (fact[i-1] * i) % MOD
    
    def mod_inverse(x):
        return pow(x, MOD-2, MOD)
    
    def permutation(n, r):
        if r > n or r < 0 or n >= max_fact:
            return 0
        return fact[n] * mod_inverse(fact[n-r]) % MOD
    
    def combination(n, r):
        if r > n or r < 0 or n >= max_fact:
            return 0
        return fact[n] * mod_inverse(fact[r]) % MOD * mod_inverse(fact[n-r]) % MOD
    
    # Total permutations
    if max_val < max_fact:
        total_perms = permutation(max_val, n)
    else:
        # For very large max_val, use different approach
        total_perms = 1
        for i in range(n):
            total_perms = (total_perms * (max_val - i)) % MOD
    
    # Permutations with Nim-sum = 0
    if n == 1:
        zero_nim_perms = 0  # Single positive value never has XOR = 0
    elif n == 2:
        # For n=2: XOR of two distinct positive values is never 0
        # (since a ⊕ b = 0 implies a = b, but we need distinct values)
        zero_nim_perms = 0
    else:
        # For n > 2: Use the mathematical property
        # Number of combinations with XOR = 0 is C(2^(m-1) - 1, n-1)
        if m == 1:
            zero_nim_perms = 0
        else:
            half_max = (1 << (m-1)) - 1
            if half_max >= max_fact or n-1 > half_max:
                zero_nim_combos = 0
            else:
                zero_nim_combos = combination(half_max, n-1)
            
            # Each combination gives n! permutations
            if n < max_fact:
                zero_nim_perms = (zero_nim_combos * fact[n]) % MOD
            else:
                zero_nim_perms = 0
    
    result = (total_perms - zero_nim_perms + MOD) % MOD
    print(result)

if __name__ == "__main__":
    solve()

File: datasets/C++/Claude/Algorithm/hard5.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1000000007;
const int MAXN = 2000000;

long long fact[MAXN];
bool fact_computed = false;

long long power(long long base, long long exp, long long mod) {
    long long result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) {
            result = (result * base) % mod;
        }
        base = (base * base) % mod;
        exp >>= 1;
    }
    return result;
}

long long modInverse(long long a) {
    return power(a, MOD - 2, MOD);
}

void computeFactorials() {
    if (fact_computed) return;
    fact[0] = 1;
    for (int i = 1; i < MAXN; i++) {
        fact[i] = (fact[i-1] * (long long)i) % MOD;
    }
    fact_computed = true;
}

long long permutation(int n, int r) {
    if (r > n || r < 0 || n >= MAXN) return 0;
    return (fact[n] * modInverse(fact[n-r])) % MOD;
}

long long combination(int n, int r) {
    if (r > n || r < 0 || n >= MAXN) return 0;
    return (fact[n] * modInverse(fact[r]) % MOD * modInverse(fact[n-r])) % MOD;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    cin >> n >> m;
    
    long long max_val = (1LL << m) - 1;  // 2^m - 1
    
    if (n > max_val) {
        cout << 0 << endl;
        return 0;
    }
    
    computeFactorials();
    
    // Total permutations
    long long total_perms;
    if (max_val < MAXN) {
        total_perms = permutation(max_val, n);
    } else {
        total_perms = 1;
        for (int i = 0; i < n; i++) {
            total_perms = (total_perms * ((max_val - i) % MOD)) % MOD;
        }
    }
    
    // Permutations with Nim-sum = 0
    long long zero_nim_perms = 0;
    
    if (n == 1 || n == 2) {
        zero_nim_perms = 0;  // XOR of distinct positive values can't be 0 for n=1,2
    } else {
        if (m > 1) {
            long long half_max = (1LL << (m-1)) - 1;
            if (half_max < MAXN && n-1 <= half_max) {
                long long zero_nim_combos = combination(half_max, n-1);
                if (n < MAXN) {
                    zero_nim_perms = (zero_nim_combos * fact[n]) % MOD;
                }
            }
        }
    }
    
    long long result = (total_perms - zero_nim_perms + MOD) % MOD;
    cout << result << endl;
    
    return 0;
}